python/ray/_raylet.pyx

```diff
@@ -404,6 +417,259 @@ cdef class RayletClient:
+    cdef:
+        c_vector[shared_ptr[CRayObject]] by_reference_objects
+
+    results = []
+    by_reference_ids = []
+    by_reference_indices = []
+    for i in range(args.size()):
+        # Passed by value.
+        if arg_reference_ids[i].IsNil():
+            data = Buffer.make(args[i].get().GetData())
+            if (args[i].get().HasMetadata()
+                and Buffer.make(
+                    args[i].get().GetMetadata()).to_pybytes()
+                    == RAW_BUFFER_METADATA):
+                results.append(data)
+            else:
+                results.append(pickle.loads(data.to_pybytes()))
+        # Passed by reference.
+        else:
+            by_reference_ids.append(
+                ObjectID(arg_reference_ids[i].Binary()))
+            by_reference_indices.append(i)
+            by_reference_objects.push_back(args[i])
+            results.append(None)
+
+    data_metadata_pairs = RayObjectsToDataMetadataPairs(
+        by_reference_objects)
+    for i, result in enumerate(
+        ray.worker.global_worker.deserialize_objects(
+            data_metadata_pairs, by_reference_ids)):
+        results[by_reference_indices[i]] = result
+
+    return results
+
+cdef _check_worker_state(worker, CTaskType task_type, JobID job_id):
+    assert worker.current_task_id.is_nil()
+    assert worker.task_context.task_index == 0
+    assert worker.task_context.put_index == 1
+
+    # If this worker is not an actor, check that `current_job_id`
+    # was reset when the worker finished the previous task.
+    if <int>task_type in [<int>TASK_TYPE_NORMAL_TASK,
+                          <int>TASK_TYPE_ACTOR_CREATION_TASK]:
+        assert worker.current_job_id.is_nil()
+        # Set the driver ID of the current running task. This is
+        # needed so that if the task throws an exception, we propagate
+        # the error message to the correct driver.
+        worker.current_job_id = job_id
+    else:
+        # If this worker is an actor, current_job_id wasn't reset.
+        # Check that current task's driver ID equals the previous
+        # one.
+        assert worker.current_job_id == job_id
+
+
+cdef _store_task_outputs(worker, return_ids, outputs):
+    for i in range(len(return_ids)):
+        return_id, output = return_ids[i], outputs[i]
+        if isinstance(output, ray.actor.ActorHandle):
+            raise Exception("Returning an actor handle from a remote "
+                            "function is not allowed).")
+        if output is ray.experimental.no_return.NoReturn:
+            if not worker.core_worker.object_exists(return_id):
+                raise RuntimeError(
+                    "Attempting to return 'ray.experimental.NoReturn' "
+                    "from a remote function, but the corresponding "
+                    "ObjectID does not exist in the local object store.")
+        else:
+            worker.put_object(return_id, output)
+
+
+cdef execute_task(
```

Comment:
Would it make sense to not expose the task type here and instead have separate execution handlers for each of the task types that the C++ worker invokes?

Reply:
I had originally implemented it this way in a previous iteration of this PR because I thought it would be simpler/cleaner, but it turned out to just add a lot of code with no clear benefit because the callbacks looked essentially the same. Would also require changing some java code which I'd rather not do in this PR. We can revisit this later once more logic is pushed into C++.